storage/innobase/xtrabackup/src/backup_mysql.cc

```diff
@@ -1074,47 +1099,86 @@ wait_for_safe_slave(MYSQL *connection)
 	}
 
 	if (strcmp(slave_sql_running, "Yes") == 0) {
+		/* Stopping slave may take significant amount of time,
+		take that into account as part of total timeout.
+		*/
+		const ssize_t before_stop_slave = (ssize_t)my_time(MY_WME);
+
+		ssize_t after_stop_slave = 0;
 		sql_thread_started = true;
 		xb_mysql_query(connection, "STOP SLAVE SQL_THREAD", false);
+
+		after_stop_slave = (ssize_t)my_time(MY_WME);
+		remaining_sleep_time -= after_stop_slave - before_stop_slave;
 	}
 
-	if (opt_safe_slave_backup_timeout > 0) {
-		n_attempts = opt_safe_slave_backup_timeout / sleep_time;
+	if (remaining_sleep_time > 0) {
+		n_attempts = remaining_sleep_time / sleep_time;
 	}
 
+retry:
 	open_temp_tables = get_open_temp_tables(connection);
 	msg_ts("Slave open temp tables: %d\n", open_temp_tables);
+	curr_slave_coordinates = get_slave_coordinates(connection);
 
-	while (open_temp_tables && n_attempts--) {
+	while (open_temp_tables && n_attempts-- > 0) {
 		msg_ts("Starting slave SQL thread, waiting %d seconds, then "
 		       "checking Slave_open_temp_tables again (%d attempts "
 		       "remaining)...\n", sleep_time, n_attempts);
+		free(prev_slave_coordinates);
+		prev_slave_coordinates = curr_slave_coordinates;
+		curr_slave_coordinates = NULL;
 
 		xb_mysql_query(connection, "START SLAVE SQL_THREAD", false);
 		os_thread_sleep(sleep_time * 1000000);
-		xb_mysql_query(connection, "STOP SLAVE SQL_THREAD", false);
+
+		curr_slave_coordinates = get_slave_coordinates(connection);
+		if (prev_slave_coordinates && curr_slave_coordinates &&
+		    strcmp(prev_slave_coordinates, curr_slave_coordinates) == 0) {
+			msg_ts("Slave pos hasn't moved during wait period, "
+			       "not stopping the SQL thread.\n");
+		}
+		else {
+			msg_ts("Stopping SQL thread.\n");
+			xb_mysql_query(connection, "STOP SLAVE SQL_THREAD",
+				false);
+		}
 
 		open_temp_tables = get_open_temp_tables(connection);
 		msg_ts("Slave open temp tables: %d\n", open_temp_tables);
 	}
 
-	/* Restart the slave if it was running at start */
 	if (open_temp_tables == 0) {
-		msg_ts("Slave is safe to backup\n");
+		xb_mysql_query(connection, "STOP SLAVE SQL_THREAD", false);
+		/* We are in a race here, slave might have opened another
```

Comment:
@gl-sergei could you please take a look, is it worth re-running a check in this case?